File: neetcode/go/_022/main.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	fmt.Println(generateParenthesis(3))
	fmt.Println(generateParenthesis(1))
}

func generateParenthesis(n int) []string {
	result := []string{}

	for bit := 0; bit < 1<<(2*n); bit++ {
		kakko := []string{}
		for i := 0; i < 2*n; i++ {
			if (bit & (1 << i)) > 0 {
				kakko = append(kakko, "(")
			} else {
				kakko = append(kakko, ")")
			}
		}

		if valid(kakko) {
			result = append(result, strings.Join(kakko, ""))
		}
	}

	return result
}

func valid(kakko []string) bool {
	stack := []string{}
	for i := range kakko {
		if kakko[i] == ")" {
			if len(stack) == 0 {
				return false
			}
			if stack[len(stack)-1] == ")" {
				return false
			}

			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, kakko[i])
		}
	}

	return len(stack) == 0
}

File: neetcode/python/_022/main.py
from typing import List

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        length = 2 * n
        for bit in range(1<<length):
            parenthesis = ""
            for i in range(length):
                parenthesis += "(" if bit & 1<<i else ")"
            if self.helper(parenthesis):
                res.append(parenthesis)
        return res

    def helper(self, parenthesis: str) -> bool:
        stack = []
        for p in parenthesis:
            if len(stack) and p == ")" and stack[-1] == "(":
                stack.pop()
            else:
                stack.append(p)
        return len(stack) == 0

print(Solution().generateParenthesis(3))
print(Solution().generateParenthesis(1))